#146 Reactive streams implementation for postgresql

antonzherdev opened the pull request:
I implemented reactive streams support for postgresql and connection pools. The function is not implemented for mysql yet. To achieve this I had to refactor the code, so it is quite difficult to review the changes. Sorry about that. Do not hesitate to ask me about changes if something is unclear.

I changed ArrayRowData to immutable in order to prevent copy of data. It is not a part of implementation but it was easier to use this immutable version of the class, and I suppose it is good from performance perspective. 

Also I faced with a thread-safety issue in PostgreSQLConnection. It was when you have got an error and start a next query. In this case the next error could be wiped out in onReadyForQuery because error processing was in the onError function. I fixed it.

antonzherdev commented:
Kerberos tests was broken but PostgreSQLConnectionSpec had a wrong package name. Therefore these tests did not execute. I disabled these tests.

mauricio commented:
@antonzherdev thanks for the PR!

It's going to take me a while to review this. As for the tests, the Kerberos tests requires PostgreSQL 8, these tests are running correctly on travis-ci so they should not be disabled.

If you have other changes that are not part of the PR (like the row data change) it would be much better to have them as separate PRs as it makes it even harder for me to review it all, since it will be hard to understand why it was changed.

antonzherdev commented:
I sent PR at first without disabling Kerberos tests but the build on Travis failed. I am not sure, but previously PostgreSQLConnectionSpec had a wrong package name and on my environment this test did not start because of this. So probably it was not executed by Travis as well.

I understand that it is difficult to review. In fact, all the changes are related because I had to refactor code as a part of this PR to make it common. I had to create RowData in another place and I wanted to avoid the code which copies data. Therefore, I changed this. Speaking about thread-safety, ConnectionPoolSpec became unstable after my changes. I fixed it but I think the issue was before but reproduced more rarely.

mauricio commented:
@antonzherdev do you think you could break this into smaller PRs? Like doing the refactoring alone or doing it in pieces?

antonzherdev commented:
@mauricio Yes, sure. I will send you few PRs with refactoring at first. And then a PR with implementation because it depends on the refactoring.
